Add EIP-1271 tutorial to developer docs

nathanhleung opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

This PR adds a tutorial about the EIP-1271 (smart contract signature validation) standard to the developer docs, under the "Ethereum Development Tutorials" section.

I wrote this page based on some of my work as a contractor for a web3 startup; the founders of the company generously gave me permission to share some of that knowledge here after I removed company-specific information. I'm happy to share more context/proof of release over email (listed in profile) if necessary.

Let me know what you think!

## Related Issue

<!--- This project accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

This page was already written so I figured I'd open a PR right away rather than opening a preliminary issue; if this sort of content isn't a great fit for ethereum.org right now I'm happy to either (a) improve it to more closely align with site standards or (b) close this PR without merging.

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+The [EIP-1271](https://eips.ethereum.org/EIPS/eip-1271) standard allows smart contracts to verify signatures.
+
+In this tutorial, we give an overview of digital signatures, EIP-1271's background, and [Gnosis Safe](https://gnosis-safe.io/)'s specific implementation of EIP-1271. All together, this can serve as a starting point for implementing EIP-1271 in your own contracts.
```
```suggestion
In this tutorial, we give an overview of digital signatures, EIP-1271's background, and [Safe](https://app.safe.global/)'s specific implementation of EIP-1271. All together, this can serve as a starting point for implementing EIP-1271 in your own contracts.
```

I believe they rebranded now from Gnosis Safe to Safe. Just updating name + url.

nathanhleung replied:
Ah yes I think you're right — I can go thru and make this change throughout

nathanhleung replied:
Might still be useful to add a parenthetical (previously Gnosis Safe) like they do on their website

<img width="662" alt="Screenshot 2023-03-01 at 13 34 56" src="https://user-images.githubusercontent.com/10874225/222270044-9a34ac6f-255d-4cd9-8e15-fd3bf55a65f7.png">

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+In this context, a signature (more precisely, a “digital signature”) is a message + some sort of proof that the message came from a specific person/sender/address.
+
+For instance, a digital signature might look like this (usually combined into some sort of data structure):
```
```suggestion
For instance, a digital signature might look like this:
```
I think this cleans this up a bit

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+## What is a signature?
+
+In this context, a signature (more precisely, a “digital signature”) is a message + some sort of proof that the message came from a specific person/sender/address.
```
```suggestion
In this context, a signature (more precisely, a “digital signature”) is a message plus some sort of proof that the message came from a specific person/sender/address.
```

Nitpick, but I think we should just use the word here

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+In order to create a digital signature for use on Ethereum-based blockchains, you generally need a secret private key which no one else knows. This is what makes your signature, yours (no one else can create the same signature without knowledge of the secret key).
+
+Your Ethereum wallet (i.e. your externally-owned account/EOA) has a private key associated with it, and this is the private key that’s typically used when a website or dapp asks you for a signature (e.g. for “Log in with Ethereum”).
```
```suggestion
Your Ethereum account (i.e. your externally-owned account/EOA) has a private key associated with it, and this is the private key that’s typically used when a website or dapp asks you for a signature (e.g. for “Log in with Ethereum”).
```

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+Your Ethereum wallet (i.e. your externally-owned account/EOA) has a private key associated with it, and this is the private key that’s typically used when a website or dapp asks you for a signature (e.g. for “Log in with Ethereum”).
+
+An app can [verify a signature](https://docs.alchemy.com/docs/how-to-verify-a-message-signature-on-ethereum) you create using a third-party library like ethers.js [without knowing your private key](https://en.wikipedia.org/wiki/Public-key_cryptography) (because of math) and be confident (again, [because of math](https://crypto.stackexchange.com/questions/48816/discrete-logarithm-problem-of-ecc))  ...
```
```suggestion
An app can [verify a signature](https://docs.alchemy.com/docs/how-to-verify-a-message-signature-on-ethereum) you create using a third-party library like ethers.js [without knowing your private key](https://en.wikipedia.org/wiki/Public-key_cryptography) and be confident that _you_ were the one that created the signature.
```

Just a bit of cleanup. Open to pushback here.

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+An app can [verify a signature](https://docs.alchemy.com/docs/how-to-verify-a-message-signature-on-ethereum) you create using a third-party library like ethers.js [without knowing your private key](https://en.wikipedia.org/wiki/Public-key_cryptography) (because of math) and be confident (again, [because of math](https://crypto.stackexchange.com/questions/48816/discrete-logarithm-problem-of-ecc))  ...
+
+> In fact, because EOA digital signatures are just based on cryptography, they can be generated and verified **off-chain**! This is how gasless DAO voting works — instead of submitting votes on-chain, digital signatures can be created and verified off-chain using cryptographic libraries.
```
```suggestion
> In fact, because EOA digital signatures use public-key cryptography, they can be generated and verified **off-chain**! This is how gasless DAO voting works — instead of submitting votes on-chain, digital signatures can be created and verified off-chain using cryptographic libraries.
```

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+> In fact, because EOA digital signatures are just based on cryptography, they can be generated and verified **off-chain**! This is how gasless DAO voting works — instead of submitting votes on-chain, digital signatures can be created and verified off-chain using cryptographic libraries.
+
+While your personal/EOA wallet has a private key, smart contracts do not have any sort of private or secret key (so "Log in with Ethereum", etc. cannot natively work with smart contracts).
```
```suggestion
While EOA accounts have a private key, smart contract accounts do not have any sort of private or secret key (so "Log in with Ethereum", etc. cannot natively work with smart contract accounts).
```

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+Contracts can implement `isValidSignature` in many ways — the spec only doesn’t say much about the exact implementation.
+
+One notable contract which implements EIP-1271 is Gnosis Safe.
```
```suggestion
One notable contract which implements EIP-1271 is Safe.
```

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+One notable contract which implements EIP-1271 is Gnosis Safe.
+
+In Gnosis Safe’s code, `isValidSignature` [is implemented](https://github.com/safe-global/safe-contracts/blob/main/contracts/handler/CompatibilityFallbackHandler.sol) so that signatures can be created and verified in [two ways](https://ethereum.stackexchange.com/questions/122635/signing-messages-as-a-gnosis-safe-eip1271-support):
```
```suggestion
In Safe’s code, `isValidSignature` [is implemented](https://github.com/safe-global/safe-contracts/blob/main/contracts/handler/CompatibilityFallbackHandler.sol) so that signatures can be created and verified in [two ways](https://ethereum.stackexchange.com/questions/122635/signing-messages-as-a-gnosis-safe-eip1271-support):
```

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+1. On-chain messages
+   1. Creation: a safe owner creates a new safe transaction to “sign” a message, passing the message as data into the transaction. Once enough owners sign the transaction to reach the multisig threshold, the transaction is broadcast and run. In the transaction, there is a safe function called which adds the message to a list of “approved” messages.
+   2. Verification: call `isValidSignature` on the Gnosis Safe contract, and pass in the message to verify as the message parameter and [an empty value for the signature parameter](https://github.com/safe-global/safe-contracts/blob/main/contracts/handler/CompatibilityFallbackHandler.sol#L32) (i.e. `0x`). Gnosis will see that the signature parameter is empty and instead of cryptographically verify ...
```
```suggestion
   2. Verification: call `isValidSignature` on the Safe contract, and pass in the message to verify as the message parameter and [an empty value for the signature parameter](https://github.com/safe-global/safe-contracts/blob/main/contracts/handler/CompatibilityFallbackHandler.sol#L32) (i.e. `0x`). Safe will see that the signature parameter is empty and instead of cryptographically verifying the signature, it will know to just go ahead and check whether the message is on the list of “approved” messages.
```

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+   2. Verification: call `isValidSignature` on the Gnosis Safe contract, and pass in the message to verify as the message parameter and [an empty value for the signature parameter](https://github.com/safe-global/safe-contracts/blob/main/contracts/handler/CompatibilityFallbackHandler.sol#L32) (i.e. `0x`). Gnosis will see that the signature parameter is empty and instead of cryptographically verify ...
+2. Off-chain messages:
+   1. Creation: a safe owner creates a message off-chain, then gets other safe owners to sign the message each individually until there are enough signatures to overcome the multisig approval threshold.
+   2. Verification: call `isValidSignature`. In the message parameter, pass in the message to be verified. In the signature parameter, pass in each safe owner’s individual signatures all concatenated together, back-to-back. Gnosis will check that there are enough signatures to meet the threshold **and** that each signature is valid. If so, it will return a value indicating successful signature ve ...
```
```suggestion
   2. Verification: call `isValidSignature`. In the message parameter, pass in the message to be verified. In the signature parameter, pass in each safe owner’s individual signatures all concatenated together, back-to-back. Safe will check that there are enough signatures to meet the threshold **and** that each signature is valid. If so, it will return a value indicating successful signature verification.
```

corwintines commented on src/content/developers/tutorials/eip-1271-smart-contract-signatures/index.md:
```diff
@@ -0,0 +1,123 @@
+
+## How should EIP-1271 be implemented in my own contracts?
+
+The spec is very open-ended here. The Gnosis Safe implementation has some good ideas:
```
```suggestion
The spec is very open-ended here. The Safe implementation has some good ideas:
```

corwintines reviewed (changes requested):
Nice, thanks @nathanhleung!

Left some change requests I think are needed. Ill also add this to our GitHub grooming call to discuss with the core team and let you know next steps :)

nathanhleung commented:
Thanks for taking a look @corwintines ! I'll take one last look over everything later this week and then re-request review when ready

nathanhleung commented:
Thanks for the comments! I went through the entire document and changed references to "Gnosis Safe" to simply "Safe." I also added a parenthetical "(previously Gnosis Safe)" in case readers are not aware of the rebrand.

nathanhleung commented:
Hey @corwintines, wanted to bump this in case it got buried. Thanks for your comments!

corwintines reviewed (approved):
Thanks @nathanhleung
